pallets/parachain-staking/src/auto_compounding.rs

```diff
@@ -0,0 +1,288 @@
+		.iter_mut()
+		.find(|entry| entry.delegator == delegator);
+
+	let mut delegation = if let Some(delegation) = maybe_delegation {
+		delegation
+	} else {
+		delegations_config.push(DelegationAutoCompoundConfig {
+			delegator,
+			value: Percent::zero(),
+		});
+		delegations_config.last_mut().expect("cannot fail; qed")
+	};
+
+	delegation.value = value;
+}
+
+/// Removes the auto-compounding value for a delegation.
+pub fn remove_delegation_config<AccountId: Eq>(
+	delegations_config: &mut Vec<DelegationAutoCompoundConfig<AccountId>>,
+	delegator: &AccountId,
+) -> bool {
+	if let Some(index) = delegations_config
+		.iter()
+		.position(|entry| &entry.delegator == delegator)
+	{
+		delegations_config.remove(index);
+		true
+	} else {
+		false
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Delegates and sets the auto-compounding config
+	pub(crate) fn delegate_with_auto_compound_config(
+		candidate: T::AccountId,
+		delegator: T::AccountId,
+		amount: BalanceOf<T>,
+		auto_compound: Percent,
+		candidate_delegation_count_hint: u32,
+		candidate_auto_compounding_delegation_count_hint: u32,
+		delegation_count_hint: u32,
+	) -> DispatchResultWithPostInfo {
+		// check that caller can reserve the amount before any changes to storage
+		ensure!(
+			Self::get_delegator_stakable_free_balance(&delegator) >= amount,
+			Error::<T>::InsufficientBalance
+		);
+
+		let mut delegator_state = if let Some(mut state) = <DelegatorState<T>>::get(&delegator) {
+			// delegation after first
+			ensure!(
+				amount >= T::MinDelegation::get(),
+				Error::<T>::DelegationBelowMin
+			);
+			ensure!(
+				delegation_count_hint >= state.delegations.0.len() as u32,
+				Error::<T>::TooLowDelegationCountToDelegate
+			);
+			ensure!(
+				(state.delegations.0.len() as u32) < T::MaxDelegationsPerDelegator::get(),
+				Error::<T>::ExceedMaxDelegationsPerDelegator
+			);
+			ensure!(
+				state.add_delegation(Bond {
+					owner: candidate.clone(),
+					amount
+				}),
+				Error::<T>::AlreadyDelegatedCandidate
+			);
+			state
+		} else {
+			// first delegation
+			ensure!(
+				amount >= T::MinDelegatorStk::get(),
+				Error::<T>::DelegatorBondBelowMin
+			);
+			ensure!(!Self::is_candidate(&delegator), Error::<T>::CandidateExists);
+			Delegator::new(delegator.clone(), candidate.clone(), amount)
+		};
+		let mut candidate_state =
+			<CandidateInfo<T>>::get(&candidate).ok_or(Error::<T>::CandidateDNE)?;
+		ensure!(
+			candidate_delegation_count_hint >= candidate_state.delegation_count,
+			Error::<T>::TooLowCandidateDelegationCountToDelegate
+		);
+		let mut auto_compounding_state = <AutoCompoundingDelegations<T>>::get(&candidate);
+		ensure!(
+			auto_compounding_state.len() as u32 <= candidate_auto_compounding_delegation_count_hint,
+			<Error<T>>::TooLowCandidateAutoCompoundingDelegationCountToDelegate,
+		);
+
+		// add delegation to candidate
+		let (delegator_position, less_total_staked) = candidate_state.add_delegation::<T>(
```

Comment:
I dont understand why we have almost identical `delegate` and `delegate_with_auto_compound` functions. Couldnt we just join them and do the auto-compound checks and insertions only if value>0¿?

Reply:
I think we wish to deprecate `delegate` in favor of `delegate_with_auto_compound`. Adding them isn't trivial without some refactoring since calling one function after the other will not reset storage & event if the 2nd function fails on validation. If code duplication is the only problem, I'd need to split the validation of `delegate` + modify local fallible structs, then call validation of `set_auto_compound` then perform storage operations - I chose to copy it directly since the other delegate should be going away.

Reply:
It's better to refactor and pool the code, because it's likely that we'll still be maintaining the `delegate` call for a while, while the eco-system migrates.